metadata_please/source_checkout_ast.py

```diff
@@ -0,0 +1,121 @@
+"""
+Reads static values from setup.py when they are simple enough.
+
+With the goal of just getting dependencies, and returning a clear error if we don't understand, this has a simpler
+
+This only reads ~50% of current setup.py, vs dowsing which is more like 80%.
+
+I experimented with a more complex version of this in
+[dowsing](https://github.com/python-packaging/dowsing/) with a goal of 100%
+coverage of open source
+"""
+
+import ast
+
+
+# Copied from orig-index
+class ShortCircuitingVisitor(ast.NodeVisitor):
+    """
+    This visitor behaves more like libcst.CSTVisitor in that a visit_ method
+    can return true or false to specify whether children get visited, and the
+    visiting of children is not the responsibility of the visit_ method.
+    """
+
+    def visit(self, node):
+        method = "visit_" + node.__class__.__name__
+        visitor = getattr(self, method, self.generic_visit)
+        rv = visitor(node)
+        if rv:
+            self.visit_children(node)
+
+    def visit_children(self, node):
+        for field, value in ast.iter_fields(node):
+            if isinstance(value, list):
+                for item in value:
+                    if isinstance(item, ast.AST):
+                        self.visit(item)
+            elif isinstance(value, ast.AST):
+                self.visit(value)
+
+    def generic_visit(self, node) -> bool:
+        return True
+
+
+class QualifiedNameSaver(ShortCircuitingVisitor):
```

Comment:
Note: currently doesn't use the short-circuiting feature for anything but skipping `func(setup(...))` by accident.